### PR TITLE
Preserve dictionary encoding when decoding parquet into Arrow arrays, 60x perf improvement (#171)

### LINKED ISSUE #171: Implement returning dictionary arrays from parquet reader
*Note*: migrated from original JIRA: https://issues.apache.org/jira/browse/ARROW-11410

Currently the Rust parquet reader returns a regular array (e.g. string array) even when the column is dictionary encoded in the parquet file.

If the parquet reader had the ability to return dictionary arrays for dictionary encoded columns this would bring many benefits such as:
 * faster reading of dictionary encoded columns from parquet (as no conversion/expansion into a regular array would be necessary)
 * more efficient memory use as the dictionary array would use less memory when loaded in memory
 * faster filtering operations as SIMD can be used to filter over the numeric keys of a dictionary string array instead of comparing string values in a string array

[~nevime] , [~alamb]  let me know what you think

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #171.

# Rationale for this change
 
See ticket

# What changes are included in this PR?

Adds dictionary preservation when decoding dictionary encoded parquet data into dictionary encoded arrays.

# Are there any user-facing changes?

No

### CODE REVIEW

File: parquet/src/arrow/bit_util.rs [hunk 1]
```diff
@@ -0,0 +1,95 @@
+// Licensed to the Apache Software Foundation (ASF) under one
```
Comment (tustvold) [inline parquet/src/arrow/bit_util.rs:1]: Moved from definition_levels.rs

File: parquet/src/arrow/record_reader/definition_levels.rs [hunk 2]
```diff
@@ -351,39 +346,6 @@ impl PackedDecoder {
     }
 }
 
-/// Counts the number of set bits in the provided range
```
Comment (tustvold) [inline parquet/src/arrow/record_reader/definition_levels.rs:354]: Moved into `crate::arrow::bit_util`

File: parquet/src/arrow/array_reader/byte_array.rs [hunk 3]
```diff
@@ -668,15 +668,9 @@ mod tests {
             assert_eq!(decoder.read(&mut output, 4..8).unwrap(), 0);
 
             let valid = vec![false, false, true, true, false, true, true, false, false];
-            let rev_position_iter = valid
```
Comment (tustvold) [inline parquet/src/arrow/array_reader/byte_array.rs:671]: This is an example of how changing pad_nulls to take a packed bitmask makes the tests less verbose

File: parquet/src/arrow/array_reader/byte_array.rs [hunk 4]
```diff
@@ -579,69 +579,18 @@ impl ByteArrayDecoderDictionary {
...
-        let mut encoder =
-            get_encoder::<ByteArrayType>(descriptor, encoding, mem_tracker).unwrap();
-
-        encoder.put(data).unwrap();
-        encoder.flush_buffer().unwrap()
-    }
 
     #[test]
     fn test_byte_array_decoder() {
-        let data: Vec<_> = vec!["hello", "world", "a", "b"]
```
Comment (tustvold) [inline parquet/src/arrow/array_reader/byte_array.rs:617]: This code is extracted out into test_utils so that it can be reused

File: parquet/src/arrow/array_reader/byte_array.rs [hunk 5]
```diff
@@ -579,69 +579,18 @@ impl ByteArrayDecoderDictionary {
...
     use crate::arrow::record_reader::buffer::ValuesBuffer;
-    use crate::basic::Type as PhysicalType;
-    use crate::data_type::{ByteArray, ByteArrayType};
-    use crate::encodings::encoding::{get_encoder, DictEncoder, Encoder};
-    use crate::schema::types::{ColumnDescriptor, ColumnPath, Type};
-    use crate::util::memory::MemTracker;
     use arrow::array::{Array, StringArray};
-    use std::sync::Arc;
-
-    fn column() -> ColumnDescPtr {
```
Comment (tustvold) [inline parquet/src/arrow/array_reader/byte_array.rs:591]: This code is extracted out into test_utils

**Comment by tustvold** [comment]
Unfortunately I needed to move some code around in order to write tests for this functionality, so this has ended up being another rather large diff :disappointed: but I've tried to comment where this has taken place so hopefully that helps.

File: parquet/src/arrow/array_reader/byte_array_dictionary.rs [hunk 6]
```diff
@@ -0,0 +1,515 @@
...
+        None => parquet_to_arrow_field(column_desc.as_ref())?
+            .data_type()
+            .clone(),
+    };
+
+    match &data_type {
+        ArrowType::Dictionary(key_type, value_type) => {
+            make_reader! {
+                (pages, column_desc, data_type, null_mask_only) => match (key_type.as_ref(), value_type.as_ref()) {
+                    (ArrowType::UInt8, ArrowType::Binary | ArrowType::Utf8) => (u8, i32),
```
Comment (tustvold) [inline parquet/src/arrow/array_reader/byte_array_dictionary.rs:102]: Many of these aren't properly supported by the arrow compute kernels, but I figured better to be thorough

File: parquet/src/arrow/array_reader/dictionary_buffer.rs [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+                let mut builder = ArrayDataBuilder::new(data_type.clone())
+                    .len(keys.len())
+                    .add_buffer(keys.into())
+                    .add_child_data(values.as_ref().clone());
+
+                if let Some(buffer) = null_buffer {
+                    builder = builder.null_bit_buffer(buffer);
+                }
+
+                let data = match cfg!(debug_assertions) {
```
Comment (alamb) [inline parquet/src/arrow/array_reader/dictionary_buffer.rs:154]: I like this defensive style of validating data in debug builds though skipping the expensive checks during runtime

File: parquet/src/arrow/array_reader/dictionary_buffer.rs [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+use crate::column::reader::decoder::ValuesBufferSlice;
+use crate::errors::{ParquetError, Result};
+use arrow::array::{make_array, ArrayData, ArrayDataBuilder, ArrayRef, OffsetSizeTrait};
+use arrow::buffer::Buffer;
+use arrow::datatypes::{ArrowNativeType, DataType as ArrowType};
+use std::sync::Arc;
+
+/// An array of variable length byte arrays that are potentially dictionary encoded
+/// and can be converted into a corresponding [`ArrayRef`]
+pub enum DictionaryBuffer<K: ScalarValue, V: ScalarValue> {
```
Comment (alamb) [inline parquet/src/arrow/array_reader/dictionary_buffer.rs:31]: this is a clever abstraction

File: parquet/src/arrow/array_reader/dictionary_buffer.rs [hunk 9]
```diff
@@ -0,0 +1,370 @@
...
+        let split = buffer.split_off(4);
+        let null_buffer = Buffer::from_iter(valid.drain(0..4));
+        let array = split.into_array(Some(null_buffer), &dict_type).unwrap();
+        assert_eq!(array.data_type(), &dict_type);
+
+        let strings = cast(&array, &ArrowType::Utf8).unwrap();
+        let strings = strings.as_any().downcast_ref::<StringArray>().unwrap();
+        assert_eq!(
+            strings.iter().collect::<Vec<_>>(),
+            vec![None, None, Some("world"), Some("hello")]
```
Comment (alamb) [inline parquet/src/arrow/array_reader/dictionary_buffer.rs:276]: Since this was splitting off the first 4 elements (indicies `1, 0, 3, 2` and valid `f,f,t,t`) I would have expected the array produced to be `None, None, Some("a"), Some("")` so I clearly misunderstand something about how this code works

File: parquet/src/arrow/array_reader/dictionary_buffer.rs [hunk 10]
```diff
@@ -0,0 +1,370 @@
...
+        let dict_type =
+            ArrowType::Dictionary(Box::new(ArrowType::Int32), Box::new(ArrowType::Utf8));
+
+        let mut buffer = DictionaryBuffer::<i32, i32>::default();
+        let d = Arc::new(StringArray::from(vec!["", "f"]).data().clone());
+        buffer.as_keys(&d).unwrap().extend_from_slice(&[0, 2, 0]);
+
+        let err = buffer.into_array(None, &dict_type).unwrap_err().to_string();
+        assert!(
+            err.contains("dictionary key beyond bounds of dictionary: 0..2"),
```
Comment (alamb) [inline parquet/src/arrow/array_reader/dictionary_buffer.rs:354]: 👍

File: parquet/src/arrow/array_reader/dictionary_buffer.rs [hunk 11]
```diff
@@ -0,0 +1,370 @@
...
+        let d3 = Arc::new(StringArray::from(vec!["bongo"]).data().clone());
+        buffer.as_keys(&d3).unwrap().extend_from_slice(&[0, 0]);
+
+        // Cannot change dictionary as keys not empty
+        let d4 = Arc::new(StringArray::from(vec!["bananas"]).data().clone());
+        assert!(buffer.as_keys(&d4).is_none());
+    }
+
+    #[test]
+    fn test_validates_keys() {
```
Comment (alamb) [inline parquet/src/arrow/array_reader/dictionary_buffer.rs:344]: Note to other reviewers that there is a test for decoding a dictionary whose keys are too large (e.g. an index of `500` when the keys type is `Int8`) in `byte_array_dictionary.rs`

**Review by alamb — COMMENTED** [review]
I went through this pretty carefully and I think I understand the code and it makes sense to me 👍  @tustvold 

I also ran the datafusion suite (`cargo test --all`) against this branch and all the tests passed

Do you have any performance results you can share here (e.g. a benchmark of reading a dictionary encoded string column from a parquet file to an Arrow array)?

Also, I think some sort of "end to end" test is probably appropriate -- i.e. that reads a parquet file and ensures that the output columns are really dictionary encoded.


FYI @yordan-pavlov  and @nevi-me and @GregBowyer 

I plan to write a mini program to validate that some of our internally generated parquet files did indeed read back a dictionary array rather than a StringArray

**Review by alamb — COMMENTED** [review]
With an end to end test I think this PR is ready to merge, but I would like to see if others reviewers would like time to do so

**Comment by alamb** [comment]
I tried reading one of our internal parquet files with this branch and I got an internal error -- I will post a reproducer shortly

**Comment by tustvold** [comment]
Yup - looks like this barfs if it encounters an empty dictionary. Will add tests and fix tomorrow :+1:

**Comment by alamb** [comment]
> Yup - looks like this barfs if it encounters an empty dictionary. Will add tests and fix tomorrow 👍

For anyone following along, I shared the internal file with @tustvold . I also have some ideas to increase the coverage of the parquet reader (including the "end to end" test I mentioned reading out a file). I'll try and write it up (and maybe even code it) tomorrow

File: parquet/src/arrow/array_reader/dictionary_buffer.rs (hunk 9 above)
Comment (tustvold) [reply parquet/src/arrow/array_reader/dictionary_buffer.rs:276]: It reads 4 non-null values

```
[Some("world"), Some("hello"), Some("a"), Some(""), Some("b")]
```

It then pads them out based on the null mask

```
[None, None, Some("world"), Some("hello"), None, Some("a"), Some(""), Some("b")]
```

It then splits off the first four

```
[None, None, Some("world"), Some("hello")]
```

Leaving behind in the buffer

```
[None, Some("a"), Some(""), Some("b")]
```

**Comment by tustvold** [comment]
I've pushed a commit that fixes the handling of empty dictionaries. I vacillated between returning an error if trying to read data from a column with an empty dictionary, or just returning 0 values read. In the end I went with the latter as it isn't always known how many non-null values a column contains. 

If GenericColumnReader thinks there should be non-null values, as determined from the definition levels, it will return an error along the lines of "insufficient values read from column".

Whilst doing this I also changed to return an error instead of panicking if there is no dictionary page present

File: parquet/src/arrow/array_reader/byte_array_dictionary.rs [hunk 12]
```diff
@@ -0,0 +1,510 @@
...
+        /// a v1 data page
+        max_remaining_values: usize,
+    },
+    Fallback(ByteArrayDecoder),
+}
+
+/// A [`ColumnValueDecoder`] for dictionary encoded variable length byte arrays
+struct DictionaryDecoder<K, V> {
+    /// The current dictionary
+    dict: Option<Arc<ArrayData>>,
```
[the remark this answers is not in the document]
Comment (tustvold) [reply parquet/src/arrow/array_reader/byte_array_dictionary.rs]: Done - although it did result in some fun with fat pointer equality :sweat_smile:

File: parquet/src/arrow/arrow_reader.rs [hunk 13]
```diff
@@ -470,9 +454,78 @@ mod tests {
...
+        let key_types = [
+            ArrowDataType::Int16,
+            ArrowDataType::UInt16,
+            ArrowDataType::Int32,
+            ArrowDataType::UInt32,
+            ArrowDataType::Int64,
+            ArrowDataType::UInt64,
+        ];
+
+        for key in &key_types {
```
[the remark this answers is not in the document]
Comment (tustvold) [reply parquet/src/arrow/arrow_reader.rs:495]: Yeah... Tbh there may be value in splitting it out as an actual fuzz testing suite in its own right...

**Comment by yordan-pavlov** [comment]
@tustvold do you think that benchmarks should be amended to include reading of dictionary arrays?

File: parquet/src/arrow/array_reader/dictionary_buffer.rs (hunk 9 above)
Comment (alamb) [reply parquet/src/arrow/array_reader/dictionary_buffer.rs:289]: thank you for the clarification

File: parquet/src/arrow/record_reader.rs [hunk 14]
```diff
@@ -167,7 +167,13 @@ where
                 break;
             }
 
-            let batch_size = max(num_records - records_read, MIN_BATCH_SIZE);
+            // If repetition levels present, we don't know how much more to read
```
[the remark this answers is not in the document]
Comment (alamb) [reply parquet/src/arrow/record_reader.rs:170]: I recommend you add some of this explanation in as a comment -- it is very 👍

**Comment by tustvold** [comment]
Added some benchmarks

```
arrow_array_reader/read StringDictionary, dictionary encoded, mandatory, no NULLs - old                                                                             
                        time:   [1.2224 ms 1.2231 ms 1.2238 ms]
arrow_array_reader/read StringDictionary, dictionary encoded, mandatory, no NULLs - new                                                                             
                        time:   [20.440 us 20.464 us 20.490 us]
arrow_array_reader/read StringDictionary, dictionary encoded, optional, no NULLs - old                                                                             
                        time:   [1.2893 ms 1.2898 ms 1.2902 ms]
arrow_array_reader/read StringDictionary, dictionary encoded, optional, no NULLs - new                                                                             
                        time:   [30.914 us 30.923 us 30.933 us]
arrow_array_reader/read StringDictionary, dictionary encoded, optional, half NULLs - old                                                                             
                        time:   [1.1340 ms 1.1344 ms 1.1347 ms]
arrow_array_reader/read StringDictionary, dictionary encoded, optional, half NULLs - new                                                                             
                        time:   [38.225 us 38.234 us 38.243 us]
```

So a nice 60x performance improvement :smile:

File: parquet/src/arrow/array_reader/byte_array_dictionary.rs [hunk 15]
```diff
@@ -341,7 +341,9 @@ where
                     Some(keys) => {
                         // Happy path - can just copy keys
                         // Keys will be validated on conversion to arrow
-                        decoder.get_batch(keys.spare_capacity_mut(len))
+                        let keys_slice = keys.spare_capacity_mut(range.start + len);
```
Comment (tustvold) [inline parquet/src/arrow/array_reader/byte_array_dictionary.rs:344]: I need to figure out why this didn't get caught by the fuzz tests

**Comment by tustvold** [comment]
Marking as a draft whilst I work on some further tweaks

File: parquet/src/arrow/array_reader/byte_array_dictionary.rs (hunk 15 above)
Comment (tustvold) [reply parquet/src/arrow/array_reader/byte_array_dictionary.rs:344]: Fix in https://github.com/apache/arrow-rs/pull/1180/commits/bf28e160b5c3b1df923014b43f7acf39fb980dff

**Comment by tustvold** [comment]
I'm happy with the state of this PR now, there is some further refinement I'd like to do on the traits, but I would rather do that as a separate PR as this one is getting quite large

**Comment by alamb** [comment]
I have one more test I want to run locally before approving this: 
1. dump out the parquet data for some local files using this branch and another branch to make sure the same data results

Working on this now

**Review by alamb — APPROVED** [review]
I ran this branch against a bunch of production data we have here at Influx, using the https://github.com/alamb/pdump command built against this branch @ 8985aeeb2ed1f55fbb061f8771ccd390df804d42 and master at 0377aaed5ff46214359d1b8d66c27f3afd9323c3
```
-parquet = { git = "https://github.com/apache/arrow-rs.git", rev="0377aaed5ff46214359d1b8d66c27f3afd9323c3" }
+parquet = { git = "https://github.com/tustvold/arrow-rs.git", rev="8985aeeb2ed1f55fbb061f8771ccd390df804d42" }
```

I dumped each file using pdump built against the two branches and compared the results using `diff` -- which detected no differences in the output 🎉 

Epic work @tustvold

**Comment by alamb** [comment]
I plan to wait for the weekend to see if anyone else wants a chance to review this before I merge it in

**Comment by yordan-pavlov** [comment]
> So a nice 60x performance improvement

@tustvold those benchmark results look very promising indeed; the next logical question is what's the impact to end-to-end performance of queries in Data Fusion

**Comment by tustvold** [comment]
> next logical question is what's the impact to end-to-end performance of queries in Data Fusion

This is highly context dependent, queries that were previously bottlenecked by parquet will of course see improvements. A simple table scan with no predicates, for example, should see most of the raw 60x performance uplift. This sort of "query" shows up in IOx when compacting or loading data from object storage into an in-memory cache.

The story for more complex queries is a bit more WIP. Currently Datafusion's handling of dictionary encoded arrays isn't brilliant with it often fully materializing dictionaries when it shouldn't need to. https://github.com/apache/arrow-datafusion/pull/1475 and https://github.com/apache/arrow-datafusion/issues/1610 track the process of switching DataFusion to delegate comparator selection to arrow-rs, which should help to alleviate this.

TLDR at this point in time I'm focused on getting arrow-rs to a good place, with the necessary base primitives, and then I'll turn my attention to what Datafusion is doing with them :grin:

**Review by houqp — APPROVED** [review]
Nice optimization :+1:
